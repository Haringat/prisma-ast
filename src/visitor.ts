import { CstNode, IToken } from '@chevrotain/types';
import { parser } from './parser';

const BasePrismaVisitor = parser.getBaseCstVisitorConstructorWithDefaults();
export class PrismaVisitor extends BasePrismaVisitor {
  constructor() {
    super();
    this.validateVisitor();
  }

  schema(ctx: CstNode & { list: CstNode[] }) {
    const list = ctx.list && ctx.list.map(item => this.visit([item]));
    return { type: 'schema', list };
  }

  component(
    ctx: CstNode & {
      type: [IToken];
      groupName?: [IToken];
      componentName: [IToken];
      block: [CstNode];
    }
  ) {
    const [{ image: type }] = ctx.type;
    const [{ image: group }] = ctx.groupName || [{}];
    const [{ image: name }] = ctx.componentName;
    const list = this.visit(ctx.block);
    const listKey =
      type === 'datasource' || type === 'generator'
        ? 'assignments'
        : type === 'enum'
        ? 'enumerators'
        : 'properties';
    return { type, group, name, [listKey]: list };
  }

  comment(ctx: CstNode & { text: [IToken] }) {
    const [{ image: comment }] = ctx.text;
    return { type: 'comment', text: comment };
  }

  block(ctx: CstNode & { list: CstNode[] }) {
    return ctx.list && ctx.list.map(item => this.visit([item]));
  }

  property(
    ctx: CstNode & { propertyName: [IToken]; propertyValue: [CstNode] }
  ) {
    const value = this.visit(ctx.propertyValue);
    const [{ image: name }] = ctx.propertyName;
    return { type: 'property', name, value };
  }

  assignment(
    ctx: CstNode & { assignmentName: [IToken]; assignmentValue: [CstNode] }
  ) {
    const value = this.visit(ctx.assignmentValue);
    const [{ image: key }] = ctx.assignmentName;
    return { type: 'assignment', key, value };
  }

  field(
    ctx: CstNode & {
      fieldName: [IToken];
      fieldType: [CstNode];
      array: [IToken];
      optional: [IToken];
      attributeList: CstNode[];
    }
  ) {
    const fieldType = this.visit(ctx.fieldType);
    const [{ image: name }] = ctx.fieldName;
    const attributes =
      ctx.attributeList && ctx.attributeList.map(item => this.visit([item]));
    return {
      type: 'field',
      name,
      fieldType,
      array: ctx.array != null,
      optional: ctx.optional != null,
      attributes,
    };
  }

  attribute(
    ctx: CstNode & {
      modelAttribute: [IToken];
      fieldAttribute: [IToken];
      groupName: [IToken];
      attributeName: [IToken];
      attributeArg: CstNode[];
    }
  ) {
    const [{ image: name }] = ctx.attributeName;
    const [{ image: group }] = ctx.groupName || [{}];
    const args =
      ctx.attributeArg && ctx.attributeArg.map(attr => this.visit(attr));
    const kind = ctx.modelAttribute != null ? 'model' : 'field';
    return { type: 'attribute', name, kind, group, args };
  }

  attributeArg(ctx: CstNode & { value: [CstNode] }) {
    const value = this.visit(ctx.value);
    return { type: 'attributeArgument', value };
  }

  func(ctx: CstNode & { funcName: [IToken]; value: CstNode[] }) {
    const [{ image: name }] = ctx.funcName;
    const params = ctx.value && ctx.value.map(item => this.visit([item]));
    return { type: 'function', name, params };
  }

  array(ctx: CstNode & { value: CstNode[] }) {
    const args = ctx.value && ctx.value.map(item => this.visit([item]));
    return { type: 'array', args };
  }

  keyedArg(ctx: CstNode & { keyName: [IToken]; value: [CstNode] }) {
    const [{ image: key }] = ctx.keyName;
    const value = this.visit(ctx.value);
    return { type: 'keyValue', key, value };
  }

  value(ctx: CstNode & { value: [IToken] | [CstNode] }) {
    if (isToken(ctx.value)) {
      const [{ image }] = ctx.value;
      return image;
    }
    return this.visit(ctx.value);
  }

  enum(ctx: CstNode & { enumName: [IToken] }) {
    const [{ image: name }] = ctx.enumName;
    return { type: 'enumerator', name };
  }
}

function isToken(node: [IToken] | [CstNode]): node is [IToken] {
  return 'image' in node[0];
}
