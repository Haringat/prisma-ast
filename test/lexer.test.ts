import { PrismaLexer } from '../src/lexer';
import { loadFixture } from './utils';

describe('lexer', () => {
  it('can tokenize', async () => {
    const source = await loadFixture('star.prisma');
    const tokens = PrismaLexer.tokenize(source);
    expect(tokens).toMatchSnapshot();
  });
});
