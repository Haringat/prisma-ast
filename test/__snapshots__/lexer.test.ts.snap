// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lexer can tokenize 1`] = `
Object {
  "errors": Array [],
  "groups": Object {},
  "tokens": Array [
    Object {
      "endColumn": 60,
      "endLine": 1,
      "endOffset": 59,
      "image": "// an example from prisma with some atypical syntax to parse",
      "startColumn": 1,
      "startLine": 1,
      "startOffset": 0,
      "tokenType": Object {
        "CATEGORIES": Array [
          Object {
            "CATEGORIES": Array [],
            "PATTERN": /NOT_APPLICABLE/,
            "categoryMatches": Array [
              12,
              13,
            ],
            "categoryMatchesMap": Object {
              "12": true,
              "13": true,
            },
            "isParent": true,
            "name": "Comment",
            "tokenTypeIdx": 11,
          },
        ],
        "PATTERN": /\\\\/\\\\/\\\\s\\*\\(\\.\\+\\)/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "LineComment",
        "tokenTypeIdx": 13,
      },
      "tokenTypeIdx": 13,
    },
    Object {
      "endColumn": 61,
      "endLine": 1,
      "endOffset": 60,
      "image": "
",
      "startColumn": 61,
      "startLine": 1,
      "startOffset": 60,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "LABEL": "LineBreak",
        "LINE_BREAKS": true,
        "PATTERN": "
",
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "LineBreak",
        "tokenTypeIdx": 32,
      },
      "tokenTypeIdx": 32,
    },
    Object {
      "endColumn": 1,
      "endLine": 2,
      "endOffset": 61,
      "image": "
",
      "startColumn": 1,
      "startLine": 2,
      "startOffset": 61,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "LABEL": "LineBreak",
        "LINE_BREAKS": true,
        "PATTERN": "
",
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "LineBreak",
        "tokenTypeIdx": 32,
      },
      "tokenTypeIdx": 32,
    },
    Object {
      "endColumn": 5,
      "endLine": 3,
      "endOffset": 66,
      "image": "model",
      "startColumn": 1,
      "startLine": 3,
      "startOffset": 62,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "LONGER_ALT": Object {
          "CATEGORIES": Array [],
          "PATTERN": /\\[a-zA-Z\\]\\\\w\\*/,
          "categoryMatches": Array [],
          "categoryMatchesMap": Object {},
          "isParent": false,
          "name": "Identifier",
          "tokenTypeIdx": 3,
        },
        "PATTERN": /model/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "Model",
        "tokenTypeIdx": 6,
      },
      "tokenTypeIdx": 6,
    },
    Object {
      "endColumn": 10,
      "endLine": 3,
      "endOffset": 71,
      "image": "Star",
      "startColumn": 7,
      "startLine": 3,
      "startOffset": 68,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "PATTERN": /\\[a-zA-Z\\]\\\\w\\*/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "Identifier",
        "tokenTypeIdx": 3,
      },
      "tokenTypeIdx": 3,
    },
    Object {
      "endColumn": 12,
      "endLine": 3,
      "endOffset": 73,
      "image": "{",
      "startColumn": 12,
      "startLine": 3,
      "startOffset": 73,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "LABEL": "'{'",
        "PATTERN": /\\{/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "LCurly",
        "tokenTypeIdx": 20,
      },
      "tokenTypeIdx": 20,
    },
    Object {
      "endColumn": 13,
      "endLine": 3,
      "endOffset": 74,
      "image": "
",
      "startColumn": 13,
      "startLine": 3,
      "startOffset": 74,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "LABEL": "LineBreak",
        "LINE_BREAKS": true,
        "PATTERN": "
",
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "LineBreak",
        "tokenTypeIdx": 32,
      },
      "tokenTypeIdx": 32,
    },
    Object {
      "endColumn": 4,
      "endLine": 4,
      "endOffset": 78,
      "image": "id",
      "startColumn": 3,
      "startLine": 4,
      "startOffset": 77,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "PATTERN": /\\[a-zA-Z\\]\\\\w\\*/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "Identifier",
        "tokenTypeIdx": 3,
      },
      "tokenTypeIdx": 3,
    },
    Object {
      "endColumn": 14,
      "endLine": 4,
      "endOffset": 88,
      "image": "Int",
      "startColumn": 12,
      "startLine": 4,
      "startOffset": 86,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "PATTERN": /\\[a-zA-Z\\]\\\\w\\*/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "Identifier",
        "tokenTypeIdx": 3,
      },
      "tokenTypeIdx": 3,
    },
    Object {
      "endColumn": 35,
      "endLine": 4,
      "endOffset": 109,
      "image": "@",
      "startColumn": 35,
      "startLine": 4,
      "startOffset": 109,
      "tokenType": Object {
        "CATEGORIES": Array [
          Object {
            "CATEGORIES": Array [],
            "PATTERN": /NOT_APPLICABLE/,
            "categoryMatches": Array [
              15,
              16,
            ],
            "categoryMatchesMap": Object {
              "15": true,
              "16": true,
            },
            "isParent": true,
            "name": "Attribute",
            "tokenTypeIdx": 14,
          },
        ],
        "LABEL": "'@'",
        "PATTERN": /@/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "FieldAttribute",
        "tokenTypeIdx": 16,
      },
      "tokenTypeIdx": 16,
    },
    Object {
      "endColumn": 37,
      "endLine": 4,
      "endOffset": 111,
      "image": "id",
      "startColumn": 36,
      "startLine": 4,
      "startOffset": 110,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "PATTERN": /\\[a-zA-Z\\]\\\\w\\*/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "Identifier",
        "tokenTypeIdx": 3,
      },
      "tokenTypeIdx": 3,
    },
    Object {
      "endColumn": 39,
      "endLine": 4,
      "endOffset": 113,
      "image": "@",
      "startColumn": 39,
      "startLine": 4,
      "startOffset": 113,
      "tokenType": Object {
        "CATEGORIES": Array [
          Object {
            "CATEGORIES": Array [],
            "PATTERN": /NOT_APPLICABLE/,
            "categoryMatches": Array [
              15,
              16,
            ],
            "categoryMatchesMap": Object {
              "15": true,
              "16": true,
            },
            "isParent": true,
            "name": "Attribute",
            "tokenTypeIdx": 14,
          },
        ],
        "LABEL": "'@'",
        "PATTERN": /@/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "FieldAttribute",
        "tokenTypeIdx": 16,
      },
      "tokenTypeIdx": 16,
    },
    Object {
      "endColumn": 46,
      "endLine": 4,
      "endOffset": 120,
      "image": "default",
      "startColumn": 40,
      "startLine": 4,
      "startOffset": 114,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "PATTERN": /\\[a-zA-Z\\]\\\\w\\*/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "Identifier",
        "tokenTypeIdx": 3,
      },
      "tokenTypeIdx": 3,
    },
    Object {
      "endColumn": 47,
      "endLine": 4,
      "endOffset": 121,
      "image": "(",
      "startColumn": 47,
      "startLine": 4,
      "startOffset": 121,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "LABEL": "'('",
        "PATTERN": /\\\\\\(/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "LRound",
        "tokenTypeIdx": 22,
      },
      "tokenTypeIdx": 22,
    },
    Object {
      "endColumn": 60,
      "endLine": 4,
      "endOffset": 134,
      "image": "autoincrement",
      "startColumn": 48,
      "startLine": 4,
      "startOffset": 122,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "PATTERN": /\\[a-zA-Z\\]\\\\w\\*/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "Identifier",
        "tokenTypeIdx": 3,
      },
      "tokenTypeIdx": 3,
    },
    Object {
      "endColumn": 61,
      "endLine": 4,
      "endOffset": 135,
      "image": "(",
      "startColumn": 61,
      "startLine": 4,
      "startOffset": 135,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "LABEL": "'('",
        "PATTERN": /\\\\\\(/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "LRound",
        "tokenTypeIdx": 22,
      },
      "tokenTypeIdx": 22,
    },
    Object {
      "endColumn": 62,
      "endLine": 4,
      "endOffset": 136,
      "image": ")",
      "startColumn": 62,
      "startLine": 4,
      "startOffset": 136,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "LABEL": "')'",
        "PATTERN": /\\\\\\)/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "RRound",
        "tokenTypeIdx": 23,
      },
      "tokenTypeIdx": 23,
    },
    Object {
      "endColumn": 63,
      "endLine": 4,
      "endOffset": 137,
      "image": ")",
      "startColumn": 63,
      "startLine": 4,
      "startOffset": 137,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "LABEL": "')'",
        "PATTERN": /\\\\\\)/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "RRound",
        "tokenTypeIdx": 23,
      },
      "tokenTypeIdx": 23,
    },
    Object {
      "endColumn": 64,
      "endLine": 4,
      "endOffset": 138,
      "image": "
",
      "startColumn": 64,
      "startLine": 4,
      "startOffset": 138,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "LABEL": "LineBreak",
        "LINE_BREAKS": true,
        "PATTERN": "
",
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "LineBreak",
        "tokenTypeIdx": 32,
      },
      "tokenTypeIdx": 32,
    },
    Object {
      "endColumn": 10,
      "endLine": 5,
      "endOffset": 148,
      "image": "position",
      "startColumn": 3,
      "startLine": 5,
      "startOffset": 141,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "PATTERN": /\\[a-zA-Z\\]\\\\w\\*/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "Identifier",
        "tokenTypeIdx": 3,
      },
      "tokenTypeIdx": 3,
    },
    Object {
      "endColumn": 22,
      "endLine": 5,
      "endOffset": 160,
      "image": "Unsupported",
      "startColumn": 12,
      "startLine": 5,
      "startOffset": 150,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "PATTERN": /\\[a-zA-Z\\]\\\\w\\*/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "Identifier",
        "tokenTypeIdx": 3,
      },
      "tokenTypeIdx": 3,
    },
    Object {
      "endColumn": 23,
      "endLine": 5,
      "endOffset": 161,
      "image": "(",
      "startColumn": 23,
      "startLine": 5,
      "startOffset": 161,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "LABEL": "'('",
        "PATTERN": /\\\\\\(/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "LRound",
        "tokenTypeIdx": 22,
      },
      "tokenTypeIdx": 22,
    },
    Object {
      "endColumn": 31,
      "endLine": 5,
      "endOffset": 169,
      "image": "\\"circle\\"",
      "startColumn": 24,
      "startLine": 5,
      "startOffset": 162,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "PATTERN": /"\\(:\\?\\[\\^\\\\\\\\"\\\\n\\\\r\\]\\+\\|\\\\\\\\\\(:\\?\\[bfnrtv"\\\\\\\\/\\]\\|u\\[0-9a-fA-F\\]\\{4\\}\\)\\)\\*"/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "StringLiteral",
        "tokenTypeIdx": 29,
      },
      "tokenTypeIdx": 29,
    },
    Object {
      "endColumn": 32,
      "endLine": 5,
      "endOffset": 170,
      "image": ")",
      "startColumn": 32,
      "startLine": 5,
      "startOffset": 170,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "LABEL": "')'",
        "PATTERN": /\\\\\\)/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "RRound",
        "tokenTypeIdx": 23,
      },
      "tokenTypeIdx": 23,
    },
    Object {
      "endColumn": 33,
      "endLine": 5,
      "endOffset": 171,
      "image": "?",
      "startColumn": 33,
      "startLine": 5,
      "startOffset": 171,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "LABEL": "'?'",
        "PATTERN": /\\\\\\?/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "QuestionMark",
        "tokenTypeIdx": 18,
      },
      "tokenTypeIdx": 18,
    },
    Object {
      "endColumn": 34,
      "endLine": 5,
      "endOffset": 172,
      "image": "
",
      "startColumn": 34,
      "startLine": 5,
      "startOffset": 172,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "LABEL": "LineBreak",
        "LINE_BREAKS": true,
        "PATTERN": "
",
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "LineBreak",
        "tokenTypeIdx": 32,
      },
      "tokenTypeIdx": 32,
    },
    Object {
      "endColumn": 10,
      "endLine": 6,
      "endOffset": 182,
      "image": "example1",
      "startColumn": 3,
      "startLine": 6,
      "startOffset": 175,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "PATTERN": /\\[a-zA-Z\\]\\\\w\\*/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "Identifier",
        "tokenTypeIdx": 3,
      },
      "tokenTypeIdx": 3,
    },
    Object {
      "endColumn": 22,
      "endLine": 6,
      "endOffset": 194,
      "image": "Unsupported",
      "startColumn": 12,
      "startLine": 6,
      "startOffset": 184,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "PATTERN": /\\[a-zA-Z\\]\\\\w\\*/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "Identifier",
        "tokenTypeIdx": 3,
      },
      "tokenTypeIdx": 3,
    },
    Object {
      "endColumn": 23,
      "endLine": 6,
      "endOffset": 195,
      "image": "(",
      "startColumn": 23,
      "startLine": 6,
      "startOffset": 195,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "LABEL": "'('",
        "PATTERN": /\\\\\\(/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "LRound",
        "tokenTypeIdx": 22,
      },
      "tokenTypeIdx": 22,
    },
    Object {
      "endColumn": 31,
      "endLine": 6,
      "endOffset": 203,
      "image": "\\"circle\\"",
      "startColumn": 24,
      "startLine": 6,
      "startOffset": 196,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "PATTERN": /"\\(:\\?\\[\\^\\\\\\\\"\\\\n\\\\r\\]\\+\\|\\\\\\\\\\(:\\?\\[bfnrtv"\\\\\\\\/\\]\\|u\\[0-9a-fA-F\\]\\{4\\}\\)\\)\\*"/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "StringLiteral",
        "tokenTypeIdx": 29,
      },
      "tokenTypeIdx": 29,
    },
    Object {
      "endColumn": 32,
      "endLine": 6,
      "endOffset": 204,
      "image": ")",
      "startColumn": 32,
      "startLine": 6,
      "startOffset": 204,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "LABEL": "')'",
        "PATTERN": /\\\\\\)/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "RRound",
        "tokenTypeIdx": 23,
      },
      "tokenTypeIdx": 23,
    },
    Object {
      "endColumn": 33,
      "endLine": 6,
      "endOffset": 205,
      "image": "
",
      "startColumn": 33,
      "startLine": 6,
      "startOffset": 205,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "LABEL": "LineBreak",
        "LINE_BREAKS": true,
        "PATTERN": "
",
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "LineBreak",
        "tokenTypeIdx": 32,
      },
      "tokenTypeIdx": 32,
    },
    Object {
      "endColumn": 8,
      "endLine": 7,
      "endOffset": 213,
      "image": "circle",
      "startColumn": 3,
      "startLine": 7,
      "startOffset": 208,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "PATTERN": /\\[a-zA-Z\\]\\\\w\\*/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "Identifier",
        "tokenTypeIdx": 3,
      },
      "tokenTypeIdx": 3,
    },
    Object {
      "endColumn": 22,
      "endLine": 7,
      "endOffset": 227,
      "image": "Unsupported",
      "startColumn": 12,
      "startLine": 7,
      "startOffset": 217,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "PATTERN": /\\[a-zA-Z\\]\\\\w\\*/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "Identifier",
        "tokenTypeIdx": 3,
      },
      "tokenTypeIdx": 3,
    },
    Object {
      "endColumn": 23,
      "endLine": 7,
      "endOffset": 228,
      "image": "(",
      "startColumn": 23,
      "startLine": 7,
      "startOffset": 228,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "LABEL": "'('",
        "PATTERN": /\\\\\\(/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "LRound",
        "tokenTypeIdx": 22,
      },
      "tokenTypeIdx": 22,
    },
    Object {
      "endColumn": 31,
      "endLine": 7,
      "endOffset": 236,
      "image": "\\"circle\\"",
      "startColumn": 24,
      "startLine": 7,
      "startOffset": 229,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "PATTERN": /"\\(:\\?\\[\\^\\\\\\\\"\\\\n\\\\r\\]\\+\\|\\\\\\\\\\(:\\?\\[bfnrtv"\\\\\\\\/\\]\\|u\\[0-9a-fA-F\\]\\{4\\}\\)\\)\\*"/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "StringLiteral",
        "tokenTypeIdx": 29,
      },
      "tokenTypeIdx": 29,
    },
    Object {
      "endColumn": 32,
      "endLine": 7,
      "endOffset": 237,
      "image": ")",
      "startColumn": 32,
      "startLine": 7,
      "startOffset": 237,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "LABEL": "')'",
        "PATTERN": /\\\\\\)/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "RRound",
        "tokenTypeIdx": 23,
      },
      "tokenTypeIdx": 23,
    },
    Object {
      "endColumn": 33,
      "endLine": 7,
      "endOffset": 238,
      "image": "?",
      "startColumn": 33,
      "startLine": 7,
      "startOffset": 238,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "LABEL": "'?'",
        "PATTERN": /\\\\\\?/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "QuestionMark",
        "tokenTypeIdx": 18,
      },
      "tokenTypeIdx": 18,
    },
    Object {
      "endColumn": 35,
      "endLine": 7,
      "endOffset": 240,
      "image": "@",
      "startColumn": 35,
      "startLine": 7,
      "startOffset": 240,
      "tokenType": Object {
        "CATEGORIES": Array [
          Object {
            "CATEGORIES": Array [],
            "PATTERN": /NOT_APPLICABLE/,
            "categoryMatches": Array [
              15,
              16,
            ],
            "categoryMatchesMap": Object {
              "15": true,
              "16": true,
            },
            "isParent": true,
            "name": "Attribute",
            "tokenTypeIdx": 14,
          },
        ],
        "LABEL": "'@'",
        "PATTERN": /@/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "FieldAttribute",
        "tokenTypeIdx": 16,
      },
      "tokenTypeIdx": 16,
    },
    Object {
      "endColumn": 42,
      "endLine": 7,
      "endOffset": 247,
      "image": "default",
      "startColumn": 36,
      "startLine": 7,
      "startOffset": 241,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "PATTERN": /\\[a-zA-Z\\]\\\\w\\*/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "Identifier",
        "tokenTypeIdx": 3,
      },
      "tokenTypeIdx": 3,
    },
    Object {
      "endColumn": 43,
      "endLine": 7,
      "endOffset": 248,
      "image": "(",
      "startColumn": 43,
      "startLine": 7,
      "startOffset": 248,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "LABEL": "'('",
        "PATTERN": /\\\\\\(/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "LRound",
        "tokenTypeIdx": 22,
      },
      "tokenTypeIdx": 22,
    },
    Object {
      "endColumn": 54,
      "endLine": 7,
      "endOffset": 259,
      "image": "dbgenerated",
      "startColumn": 44,
      "startLine": 7,
      "startOffset": 249,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "PATTERN": /\\[a-zA-Z\\]\\\\w\\*/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "Identifier",
        "tokenTypeIdx": 3,
      },
      "tokenTypeIdx": 3,
    },
    Object {
      "endColumn": 55,
      "endLine": 7,
      "endOffset": 260,
      "image": "(",
      "startColumn": 55,
      "startLine": 7,
      "startOffset": 260,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "LABEL": "'('",
        "PATTERN": /\\\\\\(/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "LRound",
        "tokenTypeIdx": 22,
      },
      "tokenTypeIdx": 22,
    },
    Object {
      "endColumn": 78,
      "endLine": 7,
      "endOffset": 283,
      "image": "\\"'<(10,4),11>'::circle\\"",
      "startColumn": 56,
      "startLine": 7,
      "startOffset": 261,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "PATTERN": /"\\(:\\?\\[\\^\\\\\\\\"\\\\n\\\\r\\]\\+\\|\\\\\\\\\\(:\\?\\[bfnrtv"\\\\\\\\/\\]\\|u\\[0-9a-fA-F\\]\\{4\\}\\)\\)\\*"/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "StringLiteral",
        "tokenTypeIdx": 29,
      },
      "tokenTypeIdx": 29,
    },
    Object {
      "endColumn": 79,
      "endLine": 7,
      "endOffset": 284,
      "image": ")",
      "startColumn": 79,
      "startLine": 7,
      "startOffset": 284,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "LABEL": "')'",
        "PATTERN": /\\\\\\)/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "RRound",
        "tokenTypeIdx": 23,
      },
      "tokenTypeIdx": 23,
    },
    Object {
      "endColumn": 80,
      "endLine": 7,
      "endOffset": 285,
      "image": ")",
      "startColumn": 80,
      "startLine": 7,
      "startOffset": 285,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "LABEL": "')'",
        "PATTERN": /\\\\\\)/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "RRound",
        "tokenTypeIdx": 23,
      },
      "tokenTypeIdx": 23,
    },
    Object {
      "endColumn": 81,
      "endLine": 7,
      "endOffset": 286,
      "image": "
",
      "startColumn": 81,
      "startLine": 7,
      "startOffset": 286,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "LABEL": "LineBreak",
        "LINE_BREAKS": true,
        "PATTERN": "
",
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "LineBreak",
        "tokenTypeIdx": 32,
      },
      "tokenTypeIdx": 32,
    },
    Object {
      "endColumn": 1,
      "endLine": 8,
      "endOffset": 287,
      "image": "}",
      "startColumn": 1,
      "startLine": 8,
      "startOffset": 287,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "LABEL": "'}'",
        "PATTERN": /\\}/,
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "RCurly",
        "tokenTypeIdx": 21,
      },
      "tokenTypeIdx": 21,
    },
    Object {
      "endColumn": 2,
      "endLine": 8,
      "endOffset": 288,
      "image": "
",
      "startColumn": 2,
      "startLine": 8,
      "startOffset": 288,
      "tokenType": Object {
        "CATEGORIES": Array [],
        "LABEL": "LineBreak",
        "LINE_BREAKS": true,
        "PATTERN": "
",
        "categoryMatches": Array [],
        "categoryMatchesMap": Object {},
        "isParent": false,
        "name": "LineBreak",
        "tokenTypeIdx": 32,
      },
      "tokenTypeIdx": 32,
    },
  ],
}
`;
